### PR TITLE
Add factory to GetState

### PR DESCRIPTION
**Pull request by shargon**
Required for https://github.com/neo-project/neo/pull/2942

### CODE REVIEW

File: src/Neo.VM/ExecutionContext.cs [hunk 1]
```diff
@@ -146,12 +146,13 @@ public ExecutionContext Clone(int initialPosition)
         /// Gets custom data of the specified type. If the data does not exist, create a new one.
         /// </summary>
         /// <typeparam name="T">The type of data to be obtained.</typeparam>
+        /// <param name="factory">A delegate used to create the entry. If factory is null, new() will be used.</param>
         /// <returns>The custom data of the specified type.</returns>
-        public T GetState<T>() where T : class, new()
+        public T GetState<T>(Func<T>? factory) where T : class, new()
```
Comment (shargon) [inline src/Neo.VM/ExecutionContext.cs:151]:
```suggestion
        public T GetState<T>(Func<T>? factory = null) where T : class, new()
```

**Review by Liaojinghui — APPROVED** [review]

**Review by Liaojinghui — APPROVED** [review]

**Review by vncoelho — COMMENTED** [review]
I think that this PR should had been merged when the dependent one is merged as well. The other PR is not yet fully complete.

**Comment by vncoelho** [comment]
I see that you are using it now https://github.com/neo-project/neo/commit/47317ee35c10151d4141d8a8057b741acdafc044

**Comment by vncoelho** [comment]
In this case, maybe it is better to do this way to already integrate with the NuGet package.

**Pull request merged**